rust/lance-index/src/scalar/btree.rs

```diff
@@ -0,0 +1,809 @@
+        let page_lookup_file = store.open_index_file(BTREE_LOOKUP_NAME).await?;
+        let serialized_lookup = page_lookup_file.read_record_batch(0).await?;
+        Ok(Arc::new(Self::try_from_serialized(
+            serialized_lookup,
+            store,
+        )?))
+    }
+}
+
+struct BatchStats {
+    min: ScalarValue,
+    max: ScalarValue,
+    null_count: u32,
+}
+
+fn min_val(array: &Arc<dyn Array>) -> Result<ScalarValue> {
+    let mut acc = MinAccumulator::try_new(array.data_type())?;
+    acc.update_batch(&[array.clone()])?;
+    Ok(acc.evaluate()?)
+}
+
+fn max_val(array: &Arc<dyn Array>) -> Result<ScalarValue> {
+    let mut acc = MaxAccumulator::try_new(array.data_type())?;
+    acc.update_batch(&[array.clone()])?;
+    Ok(acc.evaluate()?)
+}
+
+fn analyze_batch(batch: &RecordBatch) -> Result<BatchStats> {
+    let values = batch.column(0);
+    let min = min_val(&values)?;
+    let max = max_val(&values)?;
+    Ok(BatchStats {
+        min,
+        max,
+        null_count: values.null_count() as u32,
+    })
+}
+
+#[async_trait]
+pub trait SubIndexTrainer: Send + Sync {
+    fn schema(&self) -> &Arc<Schema>;
+    async fn train(&self, batch: RecordBatch, writer: &mut dyn IndexWriter) -> Result<u64>;
+}
+
+struct EncodedBatch {
+    stats: BatchStats,
+    offset: u64,
+}
+
+async fn train_btree_page(
+    batch: RecordBatch,
+    batch_idx: u32,
+    sub_index_trainer: &dyn SubIndexTrainer,
+    writer: &mut dyn IndexWriter,
+) -> Result<EncodedBatch> {
+    let stats = analyze_batch(&batch)?;
+    sub_index_trainer.train(batch, writer).await?;
+    Ok(EncodedBatch {
+        stats,
+        offset: batch_idx as u64,
+    })
+}
+
+fn btree_stats_as_batch(stats: Vec<EncodedBatch>) -> Result<RecordBatch> {
+    let mins = ScalarValue::iter_to_array(stats.iter().map(|stat| stat.stats.min.clone()))?;
+    let maxs = ScalarValue::iter_to_array(stats.iter().map(|stat| stat.stats.max.clone()))?;
+    let null_counts = UInt32Array::from_iter_values(stats.iter().map(|stat| stat.stats.null_count));
+    let offsets = UInt64Array::from_iter_values(stats.iter().map(|stat| stat.offset));
+
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("min", mins.data_type().clone(), false),
+        Field::new("max", maxs.data_type().clone(), false),
+        Field::new("null_count", null_counts.data_type().clone(), false),
+        Field::new("idx_offset", offsets.data_type().clone(), false),
+    ]));
+
+    let columns = vec![
+        mins,
+        maxs,
+        Arc::new(null_counts) as Arc<dyn Array>,
+        Arc::new(offsets) as Arc<dyn Array>,
+    ];
+
+    Ok(RecordBatch::try_new(schema, columns)?)
+}
+
+/// Train a btree index from a stream of sorted page-size batches of values and row ids
+///
+/// Note: This is likely to change.  It is unreasonable to expect the caller to do the sorting
+/// and re-chunking into page-size batches.  This is left for simplicity as this feature is still
```

Comment:
[chunk_stream](https://github.com/lancedb/lance/blob/93f78a82419f4de0d0104cebeb7bc72cc9263fe5/rust/lance/src/dataset/chunker.rs#L111-L114) should help with the re-chunking.

Reply:
Yep :)  I am using it in the create index PR.